core/src/env/engine/on_chain/buffer.rs

```diff
@@ -29,63 +26,123 @@ impl StaticBuffer {
     pub const fn new() -> Self {
         Self {
             buffer: [0; Self::CAPACITY],
-            len: 0,
         }
     }
+}
 
-    /// Returns the current length of the static buffer.
-    pub fn len(&self) -> usize {
-        self.len
+impl core::ops::Index<core::ops::RangeFull> for StaticBuffer {
+    type Output = [u8];
+
+    fn index(&self, index: core::ops::RangeFull) -> &Self::Output {
+        core::ops::Index::index(&self.buffer[..], index)
     }
+}
 
-    /// Resizes the static buffer to the given length.
-    ///
-    /// # Panics
-    ///
-    /// Panics for lengths greater than its capacity.
-    pub fn resize(&mut self, new_len: usize) {
-        if new_len > Self::CAPACITY {
-            panic!("static buffer overflowed")
-        }
-        self.len = new_len;
+impl core::ops::IndexMut<core::ops::RangeFull> for StaticBuffer {
+    fn index_mut(&mut self, index: core::ops::RangeFull) -> &mut Self::Output {
+        core::ops::IndexMut::index_mut(&mut self.buffer[..], index)
     }
+}
 
-    /// Resets the length of the buffer to 0.
-    pub fn clear(&mut self) {
-        self.len = 0;
+struct EncodeScope<'a> {
+    buffer: &'a mut [u8],
+    len: usize,
+}
+
+impl<'a> From<&'a mut [u8]> for EncodeScope<'a> {
+    fn from(buffer: &'a mut [u8]) -> Self {
+        Self { buffer, len: 0 }
     }
 }
 
-impl scale::Output for StaticBuffer {
+impl<'a> EncodeScope<'a> {
+    /// Returns the capacity of the encoded scope.
+    pub fn capacity(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Returns the length of the encoded scope.
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns the internal mutable byte slice.
+    pub fn into_buffer(self) -> &'a mut [u8] {
+        self.buffer
+    }
+}
+
+impl<'a> scale::Output for EncodeScope<'a> {
     fn write(&mut self, bytes: &[u8]) {
-        if self.len + bytes.len() > Self::CAPACITY {
-            panic!("static buffer overflowed")
-        }
+        assert!(self.len() + bytes.len() <= self.capacity());
```

Comment:
How about retaining the message?
```suggestion
        assert!(self.len() + bytes.len() <= self.capacity(), "static buffer overflowed");
```

Reply:
I will retain them, however, they are only of use for us programmers since this is happening within the on-chain environment that has no panic output.